Tests/AsyncQueueTests/SemaphoreTests.swift

```diff
@@ -0,0 +1,126 @@
+
+    // MARK: XCTestCase
+
+    override func setUp() async throws {
+        try await super.setUp()
+
+        systemUnderTest = Semaphore()
+    }
+
+    override func tearDown() async throws {
+        try await super.tearDown()
+
+        while await systemUnderTest.isWaiting {
+            await systemUnderTest.signal()
+        }
+    }
+
+    // MARK: Behavior Tests
+
+    func test_wait_suspendsUntilEqualNumberOfSignalCalls() async {
+        actor CountingExecutor {
+            func enqueue(andCount incrementOnCompletion: Bool = false, _ task: @escaping @Sendable () async -> Void) async {
+                await withCheckedContinuation { continuation in
+                    // Re-enter the actor context but don't wait for the result.
+                    Task {
+                        // Now that we're back in the actor context, resume.
+                        continuation.resume()
+                        await task()
+                        if incrementOnCompletion {
+                            incrementTasksCompleted()
+                        }
+                    }
+                }
+            }
+
+            func execute(_ task: @Sendable () async -> Void) async {
+                await task()
+            }
+
+            var countedTasksCompleted = 0
+
+            private func incrementTasksCompleted() {
+                countedTasksCompleted += 1
+            }
+        }
+        let executor = CountingExecutor()
+        let iterationCount = 1_000
+
+        for _ in 1...iterationCount {
+            await executor.enqueue(andCount: true) {
+                let didSuspend = await self.systemUnderTest.wait()
+                XCTAssertTrue(didSuspend)
+
+                // Signal without waiting that our prior wait completed.
+                Task {
+                    await self.systemUnderTest.signal()
+                }
+            }
+        }
+
+        // Loop one fewer than iterationCount.
+        // The count will be zero each time because we have more `wait` than `signal` calls.
+        for _ in 0..<(iterationCount-1) {
+            await executor.enqueue {
+                await self.systemUnderTest.signal()
+                // Enqueue a task to check the completed task count to give the suspended tasks above time to resume (if they were to resume, which they won't).
+                await executor.enqueue {
+                    let completedCountedTasks = await executor.countedTasksCompleted
+                    XCTAssertEqual(completedCountedTasks, 0)
+                }
+            }
+        }
+
+        await executor.execute {
+            // Signal one last time, enabling all of the original `wait` calls to resume.
+            await self.systemUnderTest.signal()
+
+            for _ in 1...iterationCount {
+                // Wait for all enqueued `wait`s to have completed and signaled their completion.
+                await self.systemUnderTest.wait()
+            }
```

Comment:
Is it necessary to loop or would a single `await self.systemUnderTest.wait()` have the same result?

Reply:
We want to `wait()` for every `signal()` on line 83, so that we only continue execution once _all_ of the tasks enqueued on line 77 have completed.